### PR TITLE
GEODE-6973: Use cachelistener to synchronize typeToId with IdToType

### PR DESCRIPTION
**Pull request by gesterzhou**
Co-authored-by: Xiaojian Zhou <gzhou@pivotal.io>
        Co-authored-by: Donal Evans <doevans@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/pdx/internal/PeerTypeRegistration.java [hunk 1]
```diff
@@ -365,20 +379,32 @@ public int defineType(PdxType newType) {
...
       newType.setTypeId(id);
 
       updateIdToTypeRegion(newType);
-
-      typeToId.put(newType, id);
-
       return newType.getTypeId();
     } finally {
+      // flush the tmpTypeToId map for who introduced this new pdxType
+      if (!tmpTypeToId.isEmpty()) {
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/pdx/internal/PeerTypeRegistration.java]: Why is this finally needed?  How would the tmpTypeToId not already be empty here?

Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/pdx/internal/PeerTypeRegistration.java]: We have discussed that, it's safe to do it when the creator holds the d-lock. 
The listener will only add into tmpTypeToId map now. We use it to make sure that the distribution should all finished before flushing anything into typeToId map (they stayed in tmpTypeToId map now)

Comment (jhuynh1) [reply geode-core/src/main/java/org/apache/geode/pdx/internal/PeerTypeRegistration.java]: But wouldn't the line above (line 386) when we actually use the map, have cleared it?  Why do we need to reclear in the finally?  Why is the duplicated code needed in the finally block?  Same goes for the other method...

**Comment by gesterzhou** [comment]
I changed the map's name based on bruce's suggestion.

**Comment by gesterzhou** [comment]
To bruce: the whole idea is to use a listener to maintain the synchronization between local TypeToId map and the IdToType region, then we don't have to search the region any more to speed up creating a new pdxType. 

All the code changes are focused on this and based on current code structure. We do notice that some codes can be enhanced. And a refactoring of TypeRegistry can even improve the performance. 

However, I would rather to log another geode ticket for refactoring work. 

We showed the pull request to the whole commons team in a presentation, new developers all understood.

**Comment by upthewaterspout** [comment]
@bschuchardt - To make your suggestion more concrete - would it help to move pendingTypeToId and typeToId to a separate data structure that encapsulates that functionality?

IMO this whole pendingTypeToId business is essentially working around a limitation in our transactions. I would expect that if a CacheListener fires in Member A as a result of a transaction, then a `get` in Member B *after* that CacheListener event should see the change. That is unfortunately not the case.

**Comment by bschuchardt** [comment]
@upthewaterspout yes, that would help.  It's like we're writing to a structure but don't want to commit the changes in the current thread.  The changes need to be committed in another thread.  Encapsulating the collections in another object that requires that the correct locks are obtained before transferring from pending to committed would go a long way toward making these changes more understandable without hearing Gester's presentation on the solution.

**Comment by DonalEvans** [comment]
@upthewaterspout 

I removed those tests. You're right, they were potentially flaky and not testing any behaviour that wasn't accounted for elsewhere. As for where the sample JSON came from, it was just an edited version of an existing JSON resource used in other tests.

**Comment by gesterzhou** [comment]
The testJSON.txt is from ./geode-core/src/distributedTest/resources/org/apache/geode/pdx/jsonStrings/json4.txt. 

We can change it to dummy url and email.

**Comment by gesterzhou** [comment]
I will create an inner class to wrap the typeToId, pendingTypeToId, pendingEnumToId.

**Comment by gesterzhou** [comment]
I created an inner class to wrap all these local maps. Pls review again.

**Pull request merged**